# üõí Python-Estructuras-Python:

Ejemplos y uso de **Estructuras en Python**, bucles, pruebas de **caja negra** e cristal, tuplas, recursividad, **exepciones** y **debugging**. Para poder ver los ejemplos, correr los codigos en consola..

- Autor: **Alexander Jair Rojas Paria**
- Email: alexjair@gmail.com

# Docstring

Leer la documentaci√≥n de funciones:

\>\>help(nombre_funcion) (enter)

## **Pruebas de Caja Negra**

Las pruebas de caja negra se basan en la especificaci√≥n de la funci√≥n o el programa, aqu√≠ debemos probar sus **inputs** y validar los **outputs**. Se llama caja negra por que no necesitamos saber necesariamente los procesos internos del programa, solo contrastar sus resultados.

Estos tipos de pruebas son muy importantes para 2 tipos de test:

-   **Unit testing**: se realizan pruebas a cada uno de los m√≥dulos para determinar su correcto funcionamiento.
-   **Integration testing**: es cuando vemos que todos los m√≥dulos funcionan entre s√≠.

Es una buena pr√°ctica realizar los test antes de crear tus lineas de c√≥digo, esto es por que cualquier cambio que se realice a futuro los *test* estaran incorporados para determinar si los cambios cumplen lo esperado.

En Python existe la posibilidad de realizar *test* gracias a la libreria unittest. Puede ser que el siguiente c√≥digo no lo entiendas en su totalidad, pero en una pr√≥xima gu√≠a detallare mas el tema de *clases* en programaci√≥n. Por ahora te mostrare como se realizan estos *test*.

```javascript

"""Importamos la libreria de unittest."""

**import** unittest

"""Creamos una clase para los test, en este caso se llamara

CajaNegraTest, y como par√°metro."""

**class** **CajaNegraTest**(unittest.TestCase):

"""Definimos la funcion que generara el test."""

**def** **test_suma_dos_positivos**(self):

"Para nuestro ejemplo usaremos 2 parametros."""

num_1 = 10

num_2 = 5

"""Y dentro de la variable resultado

guardaremos lo que nos retornara la funci√≥n suma."""

resultado = suma(num_1, num_2)

"""Y para terminar definimos la variable resultado

y cual es el valor esperado."""

self.assertEqual(resultado, 15)

"""Para definir el m√≥dulo de Python escribimos lo siguiente."""

**if** \__name_\_ == '__main__':

unittest.main()
```

Luego de escribir nuestro archivo iremos a la consola y ejecutaremos el *test*

python3 caja_negra.py

E \# Obtenemos un error en nuestro test

======================================================================

ERROR: test_suma_dos_positivos (__main__.CajaNegraTest) \# Veremos aqui el test con error

\----------------------------------------------------------------------

Traceback (most recent call last):

File "caja_negra.py", line 9, **in** test_suma_dos_positivos

resultado = suma(num_1, num_2)

NameError: name 'suma' is not defined \# La funci√≥n suma no esta definida

\----------------------------------------------------------------------

Ran 1 test **in** 0.000s

FAILED (errors=1)

Como vimos en el recuadro anterior no definimos la funci√≥n suma, para ello vamos a crearla.

**import** unittest

**def** **suma**(num_1, num_2):

**return** num_1 + num_2

**class** **CajaNegraTest**(unittest.TestCase):

**def** **test_suma_dos_positivos**(self):

num_1 = 10

num_2 = 5

resultado = suma(num_1, num_2)

self.assertEqual(resultado, 15)

**if** \__name_\_ == '__main__':

unittest.main()

Ahora ejecutaremos denuevo nuestro *test* en la terminal.

python3 caja_negra.py

.

\----------------------------------------------------------------------

Ran 1 test **in** 0.000s

OK

## **\*\* 7 principios de Testing \*\* de acuerdo al libro de ISTQB.**

**1 Las pruebas muestran la presencia de defectos  
**Significa que las pruebas pueden demostrar que EXISTEN problemas, pero no que los problemas NO EXISTEN.  
El objetivo principal de llevar a cabo una prueba es para detectar defectos. Trabajando bajo la premisa de que cada producto contiene defectos de alg√∫n tipo, una prueba que revela los errores es generalmente mejor que una que no lo hace. Todas las pruebas por lo tanto, deben ser dise√±ados para revelar tantos errores como sea posible

**2 Las pruebas exhaustivas son imposibles***  
*Las pruebas exhaustivas tratan de cubrir todas las combinaciones posibles de datos en el software, a fin de garantizar que ninguna situaci√≥n puede surgir, una vez probado el software se ha liberado. Excepto en aplicaciones muy simples, el n√∫mero de combinaciones posibles de datos es demasiado alta, es m√°s eficaz y eficiente que los ingenieros de pruebas se centren en las funcionalidades de acuerdo a riesgos y prioridades.

**3 Pruebas tempranas.  
**Un producto (incluyendo los documentos, tales como la especificaci√≥n del producto) se puede probar tan pronto como se ha creado. ISTQB recomienda probar un producto tan pronto como sea posible, corregir los errores m√°s r√°pidamente posible. Los estudios han demostrado que los errores identificados al final del proceso de desarrollo por lo general cuestan m√°s para resolver.  
Por ejemplo: un error encontrado en las especificaciones puede ser bastante sencillo de solucionar. Sin embargo, si ese error se transfiere a la codificaci√≥n de software, una vez descubierto el error puede ser muy costoso y consume tiempo.

**4 Agrupamiento de Defectos  
**Los estudios sugieren que los problemas en un elemento de software tienden a agruparse en torno a un conjunto limitado de m√≥dulos o √°reas. Una vez que estas √°reas han sido identificadas, los administradores eficientes de prueba son capaces de enfocar las pruebas en las zonas sensibles, mientras que siguen buscando los errores en los m√≥dulos de software restantes. Me recuerda al 80/20.

**5 La paradoja del ‚ÄúPesticida‚Äù  
**Al igual que el sobre uso de un pesticida, un conjunto de pruebas que se utilizan repetidamente en √©l disminuir√° en su eficacia. Usando una variedad de pruebas y t√©cnicas expondr√° una serie de defectos a trav√©s de las diferentes √°reas del producto.

**6 La prueba es dependiente del contexto  
**Las mismas pruebas no se deben aplicar en todos los √°mbitos. Distintos productos de software tienen diferentes requisitos, funciones y prop√≥sitos. Una prueba dise√±ada para realizarse en un sitio web, por ejemplo, puede ser menos eficaz cuando se aplica a una aplicaci√≥n de intranet. Una prueba dise√±ada para una forma de pago con tarjeta de cr√©dito puede ser innecesariamente rigurosa si se realiza en un foro de discusi√≥n.  
En general, cuanto mayor es la probabilidad y el impacto de los da√±os causados ‚Äã‚Äãpor el software fallado, mayor es la inversi√≥n en la realizaci√≥n de pruebas de software.

**7 La falacia de ausencia de errores  
**Declarar que una prueba no ha descubierto ning√∫n error no es lo mismo que declarar que el software es ‚Äúlibre de errores‚Äù. Con el fin de garantizar que los procedimientos adecuados de software de prueba se lleva a cabo en todas las situaciones, los evaluadores deben asumir que todo el software contiene algunos (aunque disimulada) errores

## üòÉ **Pruebas de caja de cristal**

-   Se basan en el flujo del programa.
-   Prueba todos los caminos posibles de una funci√≥n. Ramificaciones, bucles for y while, recursi√≥n.
-   Sirve para *regression testing* o *mocks*. Esto es, cuando ya se liber√≥ el programa pero descubrimos un bug que queremos arreglar.
    -   *if:* Debemos probar todas las condiciones.
    -   *for* y *recursi√≥n:* Una donde no entre, otro donde entre una vez y otra donde entre varias.
    -   *while:* Donde la condici√≥n de entrada sea falsa y otra donde probemos los *break statements.*
-   Tambi√©n debemos probar todas las excepciones.

**import** unittest

**def** **es_mayor_de_edad**(edad):

**if** edad \>= 18:

**return** **True**

**else**:

**return** **False**

**class** **PruebaDeCristalTest**(unittest.TestCase):

**def** **test_es_mayor_de_edad**(self):

edad = 20

resultado = es_mayor_de_edad(edad)

self.assertEqual(resultado, **True**)

**def** **test_es_menor_de_edad**(self):

edad = 16

resultado = es_mayor_de_edad(edad)

self.assertEqual(resultado, **False**)

**if** \__name_\_ == '__main__':

unittest.ma

## **Verbosity=2 al m√©todo unittest.main**

**if** \__name_\_ == '__main__':

unittest.main(verbosity=2)

Se le puede agregar el par√°metro verbosity=2 al m√©todo unittest.main para que el resultado en la consola sea mas detallado.

## **Excepciones comunes:**

ImportError : una importaci√≥n falla;

IndexError : una lista se indexa con un n√∫mero fuera de rango;

NameError : se usa una variable desconocida ;

SyntaxError : el c√≥digo no se puede analizar correctamente

TypeError : se llama a una funci√≥n en un valor de un tipo inapropiado;

ValueError : se llama a una funci√≥n en un valor del tipo correcto, pero con un valor inapropiado

## **Manejo de excepciones**

Los manejos de excepciones son muy comunes en la programaci√≥n, no tienen nada de excepcional. Las excepciones de Python normalmente se relacionan con errores de sem√°ntica, tambi√©n podemos crear nuestras propias excepciones, pero cuando una excepci√≥n no se maneja (*unhandled exception*), el programa termina en error.

Las excepciones se manejan con los keywords: try, except, finally. Se pueden utilizar tambi√©n para *ramificar* programas.

No deben manejarse de manera silenciosa (por ejemplo, con print statements). Para crear tu propia excepci√≥n utiliza el keyword *raise*.

"""Creamos una funci√≥n en donde cada elemento de

una lista es dividida por un divisor definido"""

**def** **divide_elementos_de_lista**(lista, divisor):

"""El programa intentara realizar la divisi√≥n"""

**try**:

**return** [i / divisor **for** i **in** lista]

"""En caso de error de tipo ZeroDivisionError que

significa error al dividir en cero, el programa

ejecutara la siguiente instrucci√≥n"""

**except** ZeroDivisionError **as** e:

**return** lista

lista = list(range(10))

divisor = 0

print(divide_elementos_de_lista(lista, divisor))

import math

\#raiz cuadrada

*def* raiz_cuadrada_elementos_lista(*lista*, *multiplicador*):

try:

return [math.sqrt(i\*multiplicador) for i in lista]

except *ValueError* as e:

print (e)

return *f*'no existe la ra√≠z cuadrada de un negativo'

lista = *list*(range(10))

multiplicador = -8

print (raiz_cuadrada_elementos_lista(lista, multiplicador))

# **Excepciones y control de flujo**

Hasta ahora hemos visto como las excepciones nos permiten controlar los posibles errores que pueden ocurrir en nuestro c√≥digo. Sin embargo, dentro de la comunidad de Python tienen otro uso: control de flujo.

En este momento ya debes estar familiarizado con las estructuras de control flujo que ofrece Python (if... elif...else); entonces, ¬øpor qu√© es necesaria otra modalidad para controlar el flujo? Una raz√≥n muy espec√≠fica: el principio EAFP (*easier to ask for forgiveness than permission*, es m√°s f√°cil pedir perd√≥n que permiso, por sus siglas en ingl√©s).

El principio EAFP es un estilo de programaci√≥n com√∫n en Python en el cual se asumen llaves, √≠ndices o atributos v√°lidos y se captura la excepci√≥n si la suposici√≥n resulta ser falsa. Es importante resaltar que otros lenguajes de programaci√≥n favorecen el principio LBYL (*look before you leap*, revisa antes de saltar) en el cual el c√≥digo verifica de manera expl√≠cita las precondiciones antes de realizar llamadas.

Veamos ambos estilos:

\# Python

**def** **busca_pais**(paises, pais):

\&quot;&quot;&quot;

Paises es un diccionario. Pais es la llave.

Codigo con el principio EAFP.

\&quot;&quot;&quot;

**try**:

**return** paises[pais]

**except** KeyError:

**return** **None**

// Javascript

/\*\*

\* Paises es un objeto. Pais es la llave.

\* Codigo con el principio LBYL.

\*/

**function** **buscaPais**(paises, pais) {

**if**(!Object.keys(paises).includes(pais)) {

**return** **null**;

}

**return** paises[pais];

}

Como puedes ver, el c√≥digo de Python accede directamente a la llave y √∫nicamente si dicho acceso falla, entonces se captura la excepci√≥n y se provee el c√≥digo necesario. En el caso de JavaScript, se verifica primero que la llave exista en el objeto y √∫nicamente con posterioridad se accede.

Es importante resaltar que ambos estilos pueden utilizarse en Python, pero el estilo EAFP es mucho m√°s "pythonico".

